### PR TITLE
[Serializer] UnwrappingDenormalizer

### PR DESCRIPTION
**Pull request by nonanerz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

UnwrappingDenormalizer, registered with very high priority. Unwrapping the data if UNWRAP_PATH is provided.

Very often some APIs give nested responses in which we need only the child object. With UnwrappingDenormalizer we can get the needed object without creating unnecessary Model class that we don't really need.

Regarding to https://github.com/symfony/symfony/pull/28887 and https://github.com/symfony/symfony/pull/30894

Usage:

`$serialiser->deserialize('{"baz": {"foo": "bar", "inner": {"title": "value", "numbers": [5,3]}}}', Object::class, ['UnwrappingDenormalizer::UNWRAP_PATH' => '[baz][inner]'])`

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]: It should be possible to inject an existing instance of `PropertyAccess` here.

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]: Why? It could be an useful information for other (custom) denormalizers.

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]:
```suggestion
final class UnwrappingDenormalizer implements DenormalizerInterface, SerializerAwareInterface
```

And can't this normalizer implements https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Normalizer/CacheableSupportsMethodInterface.php ?

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (nonanerz) [reply src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]: This was done according to suggestion of @ogizanagi here https://github.com/symfony/symfony/pull/28887#issuecomment-438760973

Should i add some `"unwrapped"` key to `$context` to avoid circular reference instead?

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (maxhelias) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]: You can use the `SerializerAwareTrait` 😉

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]: Can be removed

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]:
```suggestion
    public function __construct(PropertyAccessorInterface $propertyAccessor = null)
```

File: src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/UnwrappingDenormalizer.php]:
```suggestion
        $this->propertyAccessor = $propertyAccessor ?? PropertyAccess::createPropertyAccessor();
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Review by stof — CHANGES_REQUESTED** [review]
Please also add unit tests for the denormalizer itself.

**Comment by nonanerz** [comment]
Thank you for the feedback @dunglas and @stof. Now everything should be fine.

**Review by dunglas — APPROVED** [review]

**Review by maxhelias — APPROVED** [review]

**Comment by nonanerz** [comment]
> Please also add unit tests for the denormalizer itself.

Hi. Can you please take a look? Waiting for your approval

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nonanerz.

**Pull request merged**
